iroh-resolver/src/resolver.rs

```diff
@@ -1119,229 +1138,151 @@ impl<T: ContentLoader> Resolver<T> {
-    ) -> Result<Ipld> {
-        let mut root = root;
+    ) -> Result<(IpldCodec, Ipld)> {
         let mut current = root;
+        let mut codec = codec;
 
         for part in path {
+            // TODO(ramfox): we currently add an empty string to the path.tail vec
+            // when the original path ends in a "/". This is causing errors because
+            // we attempt to index into the data using an empty string, which fails
+            if part.is_empty() {
+                continue;
+            }
             if let libipld::Ipld::Link(c) = current {
-                let new_codec: libipld::IpldCodec = c.codec().try_into()?;
-                ensure!(
-                    new_codec == codec,
-                    "can only resolve the same codec {:?} != {:?}",
-                    new_codec,
-                    codec
-                );
-
-                // resolve link and update if we have encountered a link
-                let loaded_cid = self.load_cid(&c, ctx).await?;
-                root = codec
-                    .decode(&loaded_cid.data)
-                    .map_err(|e| anyhow!("invalid dag json: {:?}", e))?;
-                current = root;
+                (codec, current) = self.load_ipld_link(c, ctx).await?;
             }
-
-            let index: libipld::ipld::IpldIndex = if let Ok(i) = part.parse::<usize>() {
-                i.into()
+            if codec == IpldCodec::DagPb {
+                current = self.get_dagpb_link(current, part)?;
             } else {
-                part.clone().into()
-            };
-
-            current = current.take(index)?;
+                let index: libipld::ipld::IpldIndex = if let Ok(i) = part.parse::<usize>() {
+                    i.into()
+                } else {
+                    part.clone().into()
+                };
+                current = current.take(index).map_err(|_| {
+                    anyhow!(
+                        "IPLD resolve error: Couldn't find part {} in path '{}'",
+                        part,
+                        path.join("/")
+                    )
+                })?;
+            }
+        }
+        if let libipld::Ipld::Link(c) = current {
+            (codec, current) = self.load_ipld_link(c, ctx).await?;
         }
 
-        Ok(current)
+        Ok((codec, current))
+    }
+
+    #[tracing::instrument(skip(self))]
+    async fn load_ipld_link(&self, cid: Cid, ctx: &mut LoaderContext) -> Result<(IpldCodec, Ipld)> {
+        let codec: libipld::IpldCodec = cid.codec().try_into()?;
+
+        // resolve link and update if we have encountered a link
+        let loaded_cid = self.load_cid(&cid, ctx).await?;
+
+        let ipld: Ipld = codec
+            .decode(&loaded_cid.data)
+            .map_err(|e| anyhow!("invalid {:?}: {:?}", codec, e))?;
+        Ok((codec, ipld))
+    }
+
+    #[tracing::instrument(skip(self, name))]
+    fn get_dagpb_link<I: Into<String>>(&self, ipld: Ipld, name: I) -> Result<Ipld> {
```

Comment:
this... seems like a lot of trouble. Should probably encode to some DagPb struct and add methods there that make this less gross. But wasn't sure about the consequences of leaving "ipld land". I'd like to posit that this is fine for now, but should be changed asap.

Reply:
libipld is not the nicest to use